main.cc

```diff
@@ -545,6 +545,84 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
     return linfo.host_id;
 }
 
+static bool raft_topology_change_enabled(const db::config& cfg) {
+    return cfg.consistent_cluster_management() &&
+        cfg.check_experimental(db::experimental_features_t::feature::CONSISTENT_TOPOLOGY_CHANGES);
+}
+
+static void initialize_system_keyspace(sharded<db::system_keyspace>& sys_ks,
+        sharded<cql3::query_processor>& qp,
+        sharded<locator::snitch_ptr>& snitch,
+        distributed<replica::database>& db,
+        sharded<locator::effective_replication_map_factory>& erm_factory,
+        const db::config& cfg,
+        sharded<gms::feature_service>& feature_service)
+{
+    supervisor::notify("starting system keyspace");
+    sys_ks.start(std::ref(qp), std::ref(db), std::ref(snitch)).get();
+    // TODO: stop()?
+
+    // Initialization of a keyspace is done by shard 0 only. For system
+    // keyspace, the procedure  will go through the hardcoded column
+    // families, and in each of them, it will load the sstables for all
+    // shards using distributed database object.
+    // Iteration through column family directory for sstable loading is
+    // done only by shard 0, so we'll no longer face race conditions as
+    // described here: https://github.com/scylladb/scylla/issues/1014
+    supervisor::notify("loading system sstables");
+    replica::distributed_loader::init_system_keyspace(sys_ks, erm_factory, db, system_table_load_phase::phase1).get();
+
+    // Re-enable previously enabled features on node startup.
+    // This should be done before commitlog starts replaying
+    // since some features affect storage.
+    feature_service.local().enable_features_on_startup(sys_ks.local(), raft_topology_change_enabled(cfg)).get();
+
+    db.local().maybe_init_schema_commitlog();
+
+    // Init schema tables only after enable_features_on_startup()
+    // because table construction consults enabled features.
+    // Needs to be before system_keyspace::setup_after_commitlog(), which writes to schema tables.
+    replica::distributed_loader::init_system_keyspace(sys_ks, erm_factory, db, system_table_load_phase::phase2).get();
+    sys_ks.local().cache_truncation_record().get();
+
+    supervisor::notify("starting schema commit log");
+
+    // Check there is no truncation record for schema tables.
+    // Needs to happen before replaying the schema commitlog, which interprets
+    // replay position in the truncation record.
+    db.local().get_tables_metadata().for_each_table([] (table_id, lw_shared_ptr<replica::table> table_ptr) {
+        if (table_ptr->schema()->ks_name() == db::schema_tables::NAME) {
+            if (table_ptr->get_truncation_record() != db_clock::time_point::min()) {
```

Comment:
A note (for followup, I suppose). The truncation records had been populated one step above. There's no need to walk the whole set of tables metadata manually filtering out schema_tables::NAME ones. We can patch the truncation records loader so that _it_ catches the schema_tables::NAME updates

Reply:
We need to rethink, whether we need this code at all, because if we follow this route we need to include in the `if` all tables that use schema commitlog, and this is not feasible, we want to retain truncates for them.

Reply:
> we want to retain truncates for them.

For which tables?

Reply:
Note, we also have this in database::truncate_table_on_all_shards():
```
    if (s->ks_name() == db::schema_tables::NAME) {
        throw std::runtime_error(format("Truncating of {}.{} is not allowed.", s->ks_name(), s->cf_name()));
    }
```